### PR TITLE
Add IONQ Devices

### PR DESCRIPTION
**Pull request by DominikVoigt**
This PR allows for the collection of Provencance Data for IONQ devices over the AWS Braket offering.

Additional changes:
 - Fix serving of swagger ui that broke due to an issue with the springdoc version compatability with the newer spring boot version
 - Changed the postgres port to the port used in the quantil docker
 - Renamed fidelity to gateErrorRate where appropriately (AWS actually returns the fidelity)

### CODE REVIEW

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+                for (String gateName : oneQubitGates) {
+                    Gate gate = new Gate();
+                    gate.setName(gateName);
+                    gate.setQpu(qpu);
+                    // Add gate to supported gates of the qubit it operates on
+                    sourceQubit.getSupportedGates().add(gate);
+                    gate.addOperatingQubit(sourceQubit);
+                    qubitRepository.save(sourceQubit);
+                    // TODO: This leads to an illegal state representation and it appears to work anyway. Not sure if we need it? (i am not familiar enough with JPA)
+                    // gateRepository.save(gate);
```
Comment (DominikVoigt) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: @wederbn Here I got an issue with the persistence of the gates  using JPA. However in my testing the model gets updated correctly but if I include the line an exception gets thrown. What am I doing wrong?

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+                        gate.setQpu(qpu);
+                        // Add gate to supported gates of the qubits it operates on; This sets the operating qubits of the gate automatically
+                        sourceQubit.getSupportedGates().add(gate);
+                        targetQubit.getSupportedGates().add(gate);
+                        gate.addOperatingQubit(sourceQubit);
+                        gate.addOperatingQubit(targetQubit);
+                        qubitRepository.save(sourceQubit);
+                        qubitRepository.save(targetQubit);
+                        // TODO: This leads to an illegal state representation and it appears to work anyway. Not sure if we need it? (i am not familiar enough with JPA)
+                        // gateRepository.save(gate);
```
Comment (DominikVoigt) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: Same as above

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+            } else {
+                logger.warn("The IONQ device json has not the expected format. Cannot find fidelity node.");
+            }
+            JsonNode timingNode = providerNode.get("timing");
+            if (timingNode != null) {
+                JsonNode t1 = timingNode.get("T1");
+                JsonNode t2 = timingNode.get("T2");
+                if (t1 != null) {
+                    // in seconds
+                    qubitCharacteristics.setT1Time(BigDecimal.valueOf(t1.asDouble()));
```
Comment (DominikVoigt) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: @wederbn Here we got the following issue: AWS returns all timing information in seconds instead of microseconds. But if I try to manipulate the scale to microseconds I run into a numeric overflow with the schema:
`A field with precision 20, scale 15 must round to an absolute value less than 10^5`
IThe issue seems to be that the default decimal scale is 15 which only leaves 5 decimal places in front of the decimal point. How should we handle this?

**Comment by DominikVoigt** [comment]
https://github.com/UST-QuAntiL/qprov/issues/12

**Comment by salmma** [comment]
Is QPU data only polled when calibration time changed? Could be a problem for AWS, as they are not really updating their timestamps related to calibrations

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSConstants.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
+/*******************************************************************************
+ * Copyright (c) 2020 the QProv contributors.
```
Comment (wederbn) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSConstants.java]: Please update the Copyright year for all changed files

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSConstants.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    // Checkout https://docs.aws.amazon.com/braket/latest/developerguide/braket-regions.html for regions
+    public static final Map<String, String> PROVIDERS = Map.of(
+            "ionq", "us-east-1",
+            "rigetti", "us-west-1",
+            "aws", "us-east-1"
+    );
+
+    public static final Map<String, Integer> QUBITS_PER_GATE = ImmutableMap.<String, Integer>builder()
+            .put("x", 1)
```
Comment (wederbn) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSConstants.java:47]: This information can not be retrieved and must be hard coded?

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/qiskit/service/QiskitServiceRequest.java [hunk 6]
```diff
@@ -17,7 +17,7 @@
  * limitations under the License.
  *******************************************************************************/
 
-package org.quantil.qprov.collector.providers.ibmq.qiskit.service;
```
Comment (wederbn) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/qiskit/service/QiskitServiceRequest.java]: Why was this moved to another package?

File: org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 7]
```diff
@@ -26,12 +26,42 @@
...
     {
-      "name": "qprov.ibmq.qiskit-service.hostname",
+      "name": "qprov.aws.token",
+      "type": "java.lang.String",
+      "description": "Token to enable access to AWS."
+    },
+    {
+      "name": "qprov.aws.secret-token",
+      "type": "java.lang.String",
+      "description": "Token to enable access to AWS."
```
Comment (wederbn) [inline org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: Please be a bit more specific to differentiate the two tokens :)

File: org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 8]
```diff
@@ -26,12 +26,42 @@
...
+    },
+    {
+      "name": "qprov.aws.execute-calibration",
+      "type": "java.lang.Boolean",
+      "description": "True if QProv should execute calibration circuits on IBMQ to determine required data, false otherwise. When set to false some functionality may not be available in the QProv API."
```
Comment (wederbn) [inline org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: AWS in the description?

Comment (wederbn) [reply org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: Applies also to the other descriptions

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+            public AWSCredentials getCredentials() {
+                return new BasicAWSCredentials(accessToken, secretAccessToken);
+            }
+
+            @Override
+            public void refresh() {
+
+            }
+        });
+        builder.setRegion("us-west-1");
```
Comment (wederbn) [inline org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: I would refactor this into a constant in case AWS changes the simulator region or provides multiple. Maybe even provide a list with one element for easy extensibility?

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java (hunk 1 above)
Comment (wederbn) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: The supportedGates has the annotation to cascade the changes in connected entities, i.e., gates in this case. This means the qubitRepository.save() operation also commits the newly created gate to the DB and assigns it a corresponding ID. gateRepository.save(gate); tries to add the gate to the DB again, thus, the illegal state appears. So just storing the qubit should be enough

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java (hunk 2 above)
Comment (wederbn) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: See above

File: org.quantil.qprov.core/src/main/java/org/quantil/qprov/core/model/entities/GateCharacteristics.java [hunk 10]
```diff
@@ -60,7 +51,7 @@ public class GateCharacteristics implements Comparable<GateCharacteristics> {
     private BigDecimal gateTime;
 
     @Column(precision = Constants.BIG_DECIMAL_PRECISION, scale = Constants.BIG_DECIMAL_SCALE)
-    private BigDecimal gateFidelity;
+    private BigDecimal gateErrorRate;
```
Comment (wederbn) [inline org.quantil.qprov.core/src/main/java/org/quantil/qprov/core/model/entities/GateCharacteristics.java:54]: If we rename it the IBMQ provider must calculate the error rate from the fidelity? Or was it just a wrong name until now?

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSConstants.java (hunk 5 above)
Comment (DominikVoigt) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSConstants.java:47]: Yes, to my knowledge there is no API endpoint or way to extract that from the data provided by the endpoint we use.
The endpoint we use does not distinguish between 1Q/2Q gates and provides all of them together in a string array.

File: org.quantil.qprov.core/src/main/java/org/quantil/qprov/core/model/entities/GateCharacteristics.java (hunk 10 above)
Comment (DominikVoigt) [reply org.quantil.qprov.core/src/main/java/org/quantil/qprov/core/model/entities/GateCharacteristics.java:54]: It is only a wrong name. :)

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java (hunk 9 above)
Comment (DominikVoigt) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: I changed it into a constant. In fact all simulators are available in all regions. However the API always requires a service region so I selected one arbitrarily. I now commented that fact in the `AWSConstants` file.

File: org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 8 above)
Comment (DominikVoigt) [reply org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: I adapted it

File: org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 7 above)
Comment (DominikVoigt) [reply org.quantil.qprov.collector/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: I adapted it and tried to be more descriptive

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/qiskit/service/QiskitServiceRequest.java (hunk 6 above)
Comment (DominikVoigt) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/qiskit/service/QiskitServiceRequest.java]: I reverted it :) Thought I might also need it in the AWS case but I did not need it in the end.

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java (hunk 3 above)
Comment (wederbn) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: Marie said it should be long double as data type

**Review by salmma — CHANGES_REQUESTED** [review]
> Adding AWS ENV variables to the following docker-compose does not trigger auto collect of AWS backends.
> 
> Please also adapt the README to explain, where the AWS tokens have to be placed.

--> DONE :) 
And see also Slack messages.

**Comment by salmma** [comment]
Getting overflow error when collecting data:
`17:00:43.141 [pool-2-thread-1] DEBUG org.quantil.qprov.collector.providers.aws.AWSProvider - Updating qubit characteristics...
17:00:43.256 [pool-2-thread-1] ERROR org.hibernate.engine.jdbc.spi.SqlExceptionHelper - ERROR: numeric field overflow
  Detail: A field with precision 20, scale 15 must round to an absolute value less than 10^5.
org.springframework.dao.DataIntegrityViolationException: could not execute statement [ERROR: numeric field overflow
  Detail: A field with precision 20, scale 15 must round to an absolute value less than 10^5.] [insert into qubit_characteristics (calibration_time,qubit_database_id,readout_error,t1time,t2time,database_id) values (?,?,?,?,?,?)]; SQL [insert into qubit_characteristics (calibration_time,qubit_database_id,readout_error,t1time,t2time,database_id) values (?,?,?,?,?,?)]
`

File: org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java (hunk 3 above)
Comment (salmma) [reply org.quantil.qprov.collector/src/main/java/org/quantil/qprov/collector/providers/aws/AWSProvider.java]: **EDIT:** double would round the times and error rates - we want to be precise, thus, the assumption is to adapt the BigDecimal constants 
https://github.com/UST-QuAntiL/qprov/blob/22a303d04cbe5768550187e2441cb9272c27634e/org.quantil.qprov.core/src/main/java/org/quantil/qprov/core/Constants.java#L28
https://github.com/UST-QuAntiL/qprov/blob/22a303d04cbe5768550187e2441cb9272c27634e/org.quantil.qprov.core/src/main/java/org/quantil/qprov/core/Constants.java#L30

**Comment by salmma** [comment]
With fixed BigDecimal, MultiQubitGate Time/Error is 0

**Review by salmma — APPROVED** [review]
